pkg/agent/validations.go

```diff
@@ -63,6 +62,129 @@ type validationTrace struct {
+	currentMessage  map[string]string
+	previousStatus  map[string]string
+	previousMessage map[string]string
+}
+
+func checkValidations(cluster *models.Cluster, validationHistory map[string]*validationResultHistory, log *logrus.Logger) error {
+	if validationHistory["cluster"] == nil {
+		validationHistory["cluster"] = &validationResultHistory{
+			numFailures:     make(map[string]int),
+			seen:            make(map[string]bool),
+			currentStatus:   make(map[string]string),
+			currentMessage:  make(map[string]string),
+			previousStatus:  make(map[string]string),
+			previousMessage: make(map[string]string),
+		}
+	}
+	clusterLogPrefix := "Cluster validation: "
+	clusterValidationHistory, err := updateValidationsResultHistory(clusterLogPrefix, cluster.ValidationsInfo, validationHistory["cluster"], log)
+	if err != nil {
+		return err
+	}
+	validationHistory["cluster"] = clusterValidationHistory
+
+	for _, h := range cluster.Hosts {
+		if validationHistory[h.RequestedHostname] == nil {
+			validationHistory[h.RequestedHostname] = &validationResultHistory{
+				numFailures:     make(map[string]int),
+				seen:            make(map[string]bool),
+				currentStatus:   make(map[string]string),
+				currentMessage:  make(map[string]string),
+				previousStatus:  make(map[string]string),
+				previousMessage: make(map[string]string),
+			}
+		}
+		hostLogPrefix := "Host " + h.RequestedHostname + " validation: "
+		hostValidationHistory, err := updateValidationsResultHistory(hostLogPrefix, h.ValidationsInfo, validationHistory[h.RequestedHostname], log)
+		if err != nil {
+			return err
+		}
+		validationHistory[h.RequestedHostname] = hostValidationHistory
+	}
+	return nil
+}
+
+func updateValidationsResultHistory(logPrefix string, validationsInfoString string, history *validationResultHistory, log *logrus.Logger) (*validationResultHistory, error) {
+	validationsInfo := common.ValidationsStatus{}
+	err := json.Unmarshal([]byte(validationsInfoString), &validationsInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to verify validations")
+	}
+
+	for _, validationResults := range validationsInfo {
+		for _, r := range validationResults {
+			// First time we see a validation ID
+			if _, ok := history.seen[r.ID]; !ok {
+				history.seen[r.ID] = false
+				history.numFailures[r.ID] = 0
+			}
+
+			if _, ok := history.previousStatus[r.ID]; !ok {
+				history.previousStatus[r.ID] = ""
+				history.currentStatus[r.ID] = r.Status
+			} else {
+				history.previousStatus[r.ID] = history.currentStatus[r.ID]
+				history.currentStatus[r.ID] = r.Status
+				switch r.Status {
+				case validationFailure, validationError:
+					history.numFailures[r.ID]++
+				}
+			}
+
+			if _, ok := history.previousMessage[r.ID]; !ok {
+				history.previousMessage[r.ID] = ""
+				history.currentMessage[r.ID] = r.Message
+			} else {
+				history.previousMessage[r.ID] = history.currentMessage[r.ID]
+				history.currentMessage[r.ID] = r.Message
+			}
```

Comment:
I think that `previousMessage` and `previousStatus` could be unconditionally updated if moved after ` logValidationResultHistory`